File: packages/python/src/llmwho/summary.py
"""Deterministic, layered endpoint summaries."""

from collections import Counter
from collections.abc import Iterable, Sequence


def quantile(values: Sequence[float], probability: float) -> float | None:
    """R type-7 linear quantile, with identical behavior in the Node SDK."""

    if not values:
        return None
    if not 0 <= probability <= 1:
        raise ValueError("probability must be between 0 and 1")
    ordered = sorted(float(value) for value in values)
    index = (len(ordered) - 1) * probability
    lower = int(index)
    upper = min(lower + 1, len(ordered) - 1)
    fraction = index - lower
    return ordered[lower] + (ordered[upper] - ordered[lower]) * fraction


def summarize(events: Iterable[dict]) -> dict:
    rows = list(events)
    outcomes = Counter(row["transport"]["outcome"] for row in rows)
    identities = Counter(row["identity"]["status"] for row in rows)
    durations = [float(row["transport"]["duration_ms"]) for row in rows]
    observed = Counter(
        row["identity"]["observed_model"]
        for row in rows
        if row["identity"].get("observed_model")
    )
    probe_rows = [row for row in rows if row.get("probe")]
    probe_scores = [float(row["probe"]["score"]) for row in probe_rows]
    output_sizes = [
        float(row["response"]["output_bytes"])
        for row in rows
        if row.get("response") and "output_bytes" in row["response"]
    ]
    stream_requests = sum(
        1 for row in rows if row.get("request", {}).get("stream") is True
    )
    total = len(rows)
    return {
        "events": total,
        "availability": {
            "success_rate": outcomes["success"] / total if total else None,
            "outcomes": dict(sorted(outcomes.items())),
        },
        "transport": {
            "latency_ms": {
                "p50": quantile(durations, 0.5),
                "p95": quantile(durations, 0.95),
                "p99": quantile(durations, 0.99),
            }
        },
        "identity": {
            "statuses": dict(sorted(identities.items())),
            "observed_models": dict(sorted(observed.items())),
        },
        "behavior": {
            "observed_responses": len(output_sizes),
            "stream_requests": stream_requests,
            "output_bytes_p50": quantile(output_sizes, 0.5),
        },
        "capability": {
            "probe_count": len(probe_scores),
            "mean_score": (
                sum(probe_scores) / len(probe_scores) if probe_scores else None
            ),
        },
    }

File: packages/node/src/summary.js
export function quantile(values, probability) {
  if (!values.length) return null;
  if (!(probability >= 0 && probability <= 1)) {
    throw new Error("probability must be between 0 and 1");
  }
  const ordered = values.map(Number).sort((a, b) => a - b);
  const index = (ordered.length - 1) * probability;
  const lower = Math.floor(index);
  const upper = Math.min(lower + 1, ordered.length - 1);
  const fraction = index - lower;
  return ordered[lower] + (ordered[upper] - ordered[lower]) * fraction;
}

function counts(values) {
  const result = {};
  for (const value of values) result[value] = (result[value] ?? 0) + 1;
  return Object.fromEntries(Object.entries(result).sort(([a], [b]) => a.localeCompare(b)));
}

export function summarize(events) {
  const rows = [...events];
  const outcomes = counts(rows.map((row) => row.transport.outcome));
  const statuses = counts(rows.map((row) => row.identity.status));
  const declarationStatuses = counts(
    rows.map((row) => row.model_declaration?.status ?? "missing"),
  );
  const declaredModels = counts(
    rows.map((row) => row.model_declaration?.declared_model).filter(Boolean),
  );
  const durations = rows.map((row) => Number(row.transport.duration_ms));
  const probeScores = rows.filter((row) => row.probe).map((row) => Number(row.probe.score));
  const outputSizes = rows
    .filter((row) => row.response && "output_bytes" in row.response)
    .map((row) => Number(row.response.output_bytes));
  return {
    events: rows.length,
    availability: {
      success_rate: rows.length ? (outcomes.success ?? 0) / rows.length : null,
      outcomes,
    },
    transport: {
      latency_ms: {
        p50: quantile(durations, 0.5),
        p95: quantile(durations, 0.95),
        p99: quantile(durations, 0.99),
      },
    },
    identity: { statuses },
    declarations: {
      statuses: declarationStatuses,
      declared_models: declaredModels,
    },
    behavior: {
      observed_responses: outputSizes.length,
      stream_requests: rows.filter((row) => row.request?.stream === true).length,
      output_bytes_p50: quantile(outputSizes, 0.5),
    },
    capability: {
      probe_count: probeScores.length,
      mean_score: probeScores.length
        ? probeScores.reduce((sum, value) => sum + value, 0) / probeScores.length
        : null,
    },
  };
}
